Feature/howard v09370103p01 nuid features

brucehoward-physics opened the pull request:
Updated PR that uses v09_37 series as the base. Requires changes in larpandoracontent to be useful.

PetrilloAtWork reviewed (changes requested):
I have requested one change (removal of the destructor).
Other suggestions are up to the author.
One more generic one is that the recommended name convention is using the drinking camel-case ([`[NV-05]`](https://sbnsoftware.github.io/sbn/codingconv/CodingConventionsExplained.html#NV05)), e.g. `nuTotHits` (instead of `nutothits`). I _recommend_ that convention to be adopted, but I don't understand most of the names so I don't know if they are... camelcaseable.

cjbackhouse commented:
@PetrilloAtWork we have a conflicting convention of all-lowercase field names in StandardRecord. I'm not sure if that got documented in the guidelines anywhere.

henrylay97 reviewed (approved):
All good! I know this has been discussed before so I'll just throw my two pence in here and say I preferred the capitalisation in the names of the variables. A couple are quite difficult to parse now and things like `nufspfos` or `nusps` aren't particularly obvious if you haven't got the pandora tools open in front of you!

PetrilloAtWork commented on sbnanaobj/StandardRecord/SRNuID.h:
```diff
@@ -0,0 +1,34 @@
+  class SRNuID
+  {
+  public:
+    SRNuID();
```
I confirm that, beyond the choice of syntax, the recommendation is the one @brucehoward-physics described.
We do have a guideline ([`[CF-060]`](https://github.com/SBNSoftware/SBNSoftware.github.io/blob/master/sbn/codingconv/CodingConventionsExplained.md#CF060)) _suggesting_ the use of braces. That is intended as a tie-breaker in case your mind is stuck on a tie.

Some differences between `=` and `{}` syntax (whose nickname "universal initialization" betrays its ambition) for simple type data member initialization are that the latter will complain about "narrowing" (for example if you try to assign a `double` value to a `float` variable), and that it allows explicit default initialization, for example:
```cpp
  unsigned short int fNADC {};
  short int fADC[16] {};
```
will initialize `fNADC` to `0` and all `fADC` to `0` too (with no initializer, both are undefined). You can obtain the same effect with `= 0` and `= { 0 }` respectively.
Not much difference for simple types, then. When object data members are involved, braces are a superior option (but usually equivalent to parentheses, except for noticeable exceptions like [`std::vector<int>`](https://alexsm.com/cpp-std-vector-braced-init) and for its role in [the most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse)).
I try to use braces mostly because since it's newer syntax it makes me look smarter. But my official excuse is that it's more uniform so I can usually manage to have only one type of syntax (available: `int a = 0;`, `int a (0);`, `int a { 0 };`, `int a = { 0 };`; so yeah, [`{}` is the new universal standard](https://xkcd.com/927)).

brucehoward-physics replied:
Okay - have pushed commits which add a central `kSignalingNaN` and applies it to SRSlice and SRNuID (I didn't touch the other StandardRecord objects).

I have it set up to use braces. There are some items which are not seemingly not being forced to initialize in SRSlice. Shall I add an empty `{}` here too?

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.cxx:
```diff
@@ -0,0 +1,47 @@
+////////////////////////////////////////////////////////////////////////
+// \file    SRNuID.cxx
+// \brief   SRNuID object for Neutrino ID score features (MVA inputs). This SR code copied/based on other SR objects.
+// \author  $Author: howard@fnal.gov
```
✅

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.h:
```diff
@@ -0,0 +1,34 @@
+////////////////////////////////////////////////////////////////////////
+// \file    SRNuID.h
+// \brief   SRNuID object for Neutrino ID score features (MVA inputs). This SR code copied/based on other SR objects.
+// \author  $Author: howard@fnal.gov
```
✅

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.cxx:
```diff
@@ -0,0 +1,47 @@
+
+
+  SRNuID::~SRNuID(){  }
+
```
✅

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.h:
```diff
@@ -0,0 +1,34 @@
+  {
+  public:
+    SRNuID();
+    virtual ~SRNuID();
```
✅

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.h:
```diff
@@ -0,0 +1,34 @@
+// \author  $Author: howard@fnal.gov
+////////////////////////////////////////////////////////////////////////
+#ifndef SRNUID_H
+#define SRNUID_H
```
✅

PetrilloAtWork commented on sbnanaobj/StandardRecord/SRNuID.cxx:
```diff
@@ -0,0 +1,30 @@
+
+namespace caf
+{
+  SRNuID::SRNuID()
+  {  }
+
```
Given its emptiness, the constructor can be removed, unless there is the explicit intention of forbidding [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) (which I would not forbid).

brucehoward-physics replied:
Sorry I missed your comments before finalizing my statement below. There is no intent from me here, but other Standard Record objects have empty constructors apart from variable assignment. Does this carry the same impact w.r.t. aggregate initialization? If it _does_ then I defer to @cjbackhouse and/or @FernandaPsihas to add their opinion. If not, then I guess these could be removed.

PetrilloAtWork replied:
It does.

cjbackhouse replied:
Yeah, no opinion either way about empty constructors.

brucehoward-physics commented on sbnanaobj/StandardRecord/SRNuID.cxx:
```diff
@@ -0,0 +1,47 @@
+#include "sbnanaobj/StandardRecord/SRNuID.h"
+
+#include <limits>
+#include <climits>
```
believe this is now ✅ - I mean this particular change was made but now I also made changes in SRSlice, in the process of addressing other comments.

PetrilloAtWork commented on sbnanaobj/StandardRecord/SRHelper.h:
```diff
@@ -0,0 +1,11 @@
+
+namespace caf
+{
+  const float kSignalingNaN = std::numeric_limits<float>::signaling_NaN();
```
I _recommend_ setting it `constexpr`:
```suggestion
  constexpr float kSignalingNaN = std::numeric_limits<float>::signaling_NaN();
```
---- 
This other one is more for @cjbackhouse: is it ok to tie the name `kSignalingNaN` to a `float` value? do you foresee an use for the `double` counterpart?

brucehoward-physics replied:
Updated! - Only tested in that it builds.

cjbackhouse replied:
Hmm. I do indeed think that we might one day want a double version (I normally resist doubles in the CAFs, but there are potentially legitimate reasons for one). If you assign this into a double, does the double end up NaN too, or not?
Since double will be so much rarer I think having a clunkier name like `kSignalingNaNDouble` for it would be ok.

PetrilloAtWork commented on sbnanaobj/StandardRecord/SRSlice.cxx:
```diff
@@ -6,18 +6,9 @@
-    nu_pdg(INT_MIN),
-    nu_score(std::numeric_limits<float>::signaling_NaN()),
-    self(INT_MIN)
+  SRSlice::SRSlice()
```
Same comment as for the other class: an empty default constructor impedes some language features, and unless that is intended (in which case it would be good to add a comment stating it), the empty constructor is better gone.
```suggestion
```

PetrilloAtWork reviewed (commented):
Two more language comments and a question for @cjbackhouse.
Nothing show-stopping though.

brucehoward-physics commented:
I believe I have addressed most of the comments and tested all but the very last (pretty simple) commit on a single event, confirming at least that things ran and the SRNuID values in the output CAF file are set the same as they were when I started this PR. Remaining items I am currently thinking of, please note if I'm missing something:

1. On naming - There are conflicting requests, so I'm not sure how to proceed.

2. On directly assigning values - I have set up the declaration of values in the header directly for SRSlice and SRNuID (see also point 3) for things that were being assigned in the constructor. I have a question in the thread above for @PetrilloAtWork about items not previously set in the constructor.

3. Relatedly, Chris pointed out that directly assigning values in the header would be cleaner if we had a centralized kSignalingNaN, part of the point in https://github.com/SBNSoftware/sbnanaobj/issues/24 - I have added a new header `SRHelper.h` for which any reviewer should feel free to comment. I will specifically tag @cjbackhouse and @PetrilloAtWork in this point.

PetrilloAtWork commented:
Re the naming convention, I suspect there is a naming convention in `StandardRecord` that is different from the "common" code, in which case your additions should adhere to it (as they appear to do). I just haven't received a confirmation from @cjbackhouse yet.

Re the initialization and constructors, @FernandaPsihas and @cjbackhouse can comment. My quite strong opinion is that _all_ data classes should move to initialisation with their declaration in the class definition, and no constructor nor destructor, unless there are specific reasons to do otherwise. While the presence of the virtual destructor has a visible impact on runtime (the data structures are one pointer larger) the change I am describing here on the construction has none, so it should not delay the merging. I still solicit a decision for the future.

cjbackhouse commented on sbnanaobj/StandardRecord/SRHelper.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef SBNANAOBJ_STANDARDRECORD_SRHELPER_H
```
Not sure I like `SRHelper` as a name. Maybe `SRConstants`?

brucehoward-physics replied:
Done - checked only in that it builds.

cjbackhouse commented:
All looks good to me, apart from a comment about the filename.

Which fields in SRSlice are uninitialized? Anything that's an object, or a vector of anything, will default initialize properly.

I have no problem with moving all the initialization into the headers and removing most (all?) of the constructors. I think you just volunteered to do the grunt work Gianluca :stuck_out_tongue: 

There is indeed a local naming convention in force. We should formalize it somewhere and debate some of its darker corners.

brucehoward-physics commented:
@cjbackhouse 
> Which fields in SRSlice are uninitialized? Anything that's an object, or a vector of anything, will default initialize properly.

Yeah in SRSlice they are objects or vectors, so okay cool.

brucehoward-physics commented:
Okay added `kUninitializedInt` and used it in the SRNuID.h assignments. I should say the floats could use a different other than -9999 (some could use -5 for example). for the defaults but I liked the idea of them all taking such a number. If there is push-back I can change further (especially since many slices will end up with the defaults).
